pip/configuration.py

```diff
@@ -1,126 +1,318 @@
 """Configuration management setup
+
+Some terminology:
+- name
+  As written in config files.
+- value
+  Value associated with a name
+- key
+  Name combined with it's section (section.name)
+- variant
+  A single word describing where the configuration key-value pair came from
 """
 
-import re
+import logging
 import os
-import sys
 
+from pip._vendor.six import next
 from pip._vendor.six.moves import configparser
+
+from pip.exceptions import ConfigurationError
 from pip.locations import (
-    legacy_config_file, config_basename, running_under_virtualenv,
-    site_config_files
+    new_config_file, legacy_config_file, venv_config_file,
+    running_under_virtualenv, site_config_files
 )
-from pip.utils import appdirs
+from pip.utils import ensure_dir, enum
+
+
+logger = logging.getLogger(__name__)
+
+
+# NOTE: Maybe use the optionx attribute to normalize keynames.
+def _normalize_name(name):
+    """Make a name consistent regardless of source (environment or file)
+    """
+    name = name.lower().replace('_', '-')
+    if name.startswith('--'):
+        name = name[2:]  # only prefer long opts
+    return name
+
+
+def _disassemble_key(name):
+    return name.split(".", 1)
 
 
-_environ_prefix_re = re.compile(r"^PIP_", re.I)
+def _make_key(variant, name):
```

Comment:
Since it is only used in one place and quite simple, I'd say you could ditch it.